digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java

```diff
@@ -310,59 +325,107 @@ public RestProject deleteProject(@PathParam("id") int projId)
     @Consumes("application/gzip")
     @Path("/api/projects")
     public RestProject putProject(@QueryParam("project") String name, @QueryParam("revision") String revision,
-            InputStream body)
+            InputStream body, @HeaderParam("Content-Length") long contentLength)
         throws IOException, ResourceConflictException, ResourceNotFoundException
     {
         Preconditions.checkArgument(name != null, "project= is required");
         Preconditions.checkArgument(revision != null, "revision= is required");
 
-        byte[] data = ByteStreams.toByteArray(body);
-
-        if (data.length > ARCHIVE_TOTAL_SIZE_LIMIT) {
-            // TODO throw this exception before reading all data in memory
+        if (contentLength > ARCHIVE_TOTAL_SIZE_LIMIT) {
             throw new IllegalArgumentException(String.format(ENGLISH,
                         "Size of the uploaded archive file exceeds limit (%d bytes)",
                         ARCHIVE_TOTAL_SIZE_LIMIT));
         }
 
-        ArchiveMetadata meta;
-        try (TempDir dir = tempFiles.createTempDir("push", name)) {
+        InputStream archiveStream = null;
+        try {
+            ArchiveManager.Location uploadLocation =
+                archiveManager.newArchiveLocation(getSiteId(), name, revision, contentLength);
+
+            byte[] data;
+            ArchiveManager.Upload upload;
+            if (uploadLocation.getArchiveType().equals(ArchiveType.DB)) {
+                data = ByteStreams.toByteArray(body);
```

Comment:
I created ReplicatedInputStream because it doesn't have to store the entire data in memory or on disk. It does everything in streaming way.
But hm....using disk is simpler and safer.